module/VuFindSearch/src/VuFindSearch/Backend/Solr/QueryBuilder.php

```diff
@@ -180,9 +187,119 @@ public function build(AbstractQuery $query)
+            ) {
+                continue;
+            }
+            foreach ((array)$extraParam['value'] as $value) {
+                $params->add($extraParam['param'], $value);
+            }
+        }
+
         return $params;
     }
 
+    /**
+     * Check if the conditions match for an extra parameter
+     *
+     * @param AbstractQuery $query      Search query
+     * @param array         $conditions Required conditions
+     *
+     * @return bool
+     */
+    protected function checkParamConditions(
+        AbstractQuery $query,
+        array $conditions
+    ): bool {
+        if (empty($conditions)) {
+            return true;
+        }
+        $searchTypes = $this->getSearchTypes($query);
+        foreach ($conditions as $condition) {
+            if (!is_array($condition)) {
+                continue;
+            }
+            $values = reset($condition);
+            $condition = key($condition);
+            switch ($condition) {
+            case 'SearchTypeIn':
+                if (empty(array_intersect((array)$values, $searchTypes))) {
+                    return false;
+                }
+                break;
+            case 'AllSearchTypesIn':
+                if (array_diff($searchTypes, (array)$values)) {
+                    return false;
+                }
+                break;
+            case 'SearchTypeNotIn':
+                if (!empty(array_intersect((array)$values, $searchTypes))) {
+                    return false;
+                }
+                break;
+            case 'NoDismaxParams':
+                foreach ((array)$values as $value) {
+                    if ($this->hasDismaxParamsField($searchTypes, $value)) {
+                        return false;
+                    }
+                }
+                break;
+            default:
+                throw new \Exception("Unknown parameter condition: $condition");
+            }
+        }
+        return true;
+    }
+
+    /**
+     * Check if any of the given search types has the field in DismaxParams
+     *
+     * @param array  $searchTypes Search types to check
+     * @param string $field       Field to check for
+     *
+     * @return bool
+     */
+    protected function hasDismaxParamsField(array $searchTypes, string $field): bool
+    {
+        foreach ($searchTypes as $searchType) {
+            if ($handler = $this->getSearchHandler($searchType, '')) {
+                foreach ($handler->getDismaxParams() as $param) {
+                    if (reset($param) === $field) {
+                        return true;
+                    }
+                }
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Get an array of search types used in the given search
+     *
+     * @param AbstractQuery $query Query
+     *
+     * @return array
+     */
+    protected function getSearchTypes(AbstractQuery $query): array
+    {
+        $types = [];
```

Comment:
Nitpick, but I'd move this inside the `if` since it doesn't apply outside.

Reply:
Or alternatively could we eliminate the initialization by using an array_reduce(), like:

<pre>
$callback = function ($carry, $item) {
    return array_merge($carry, $this->getSearchTypes($item));
};
return array_unique(array_reduce($query->getQueries(), $callback, []));
</pre>

Apologies if I'm misremembering how array_reduce works, but this seems like it might be a perfect use case for it.

Reply:
Thanks, that's an excellent idea. I added your example, and it works great!